lib/src/authentication/credential.dart

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// ignore_for_file: import_of_legacy_library_into_null_safe
+
+import '../exceptions.dart';
+import '../source/hosted.dart';
+
+/// Token is a structure for storing authentication credentials for third-party
+/// pub registries. A token holds registry [url], credential [kind] and [token]
+/// itself.
+///
+/// Token could be serialized into and from JSON format structured like
+/// this:
+///
+/// ```json
+/// {
+///   "url": "https://example.com/",
+///   "token": "gjrjo7Tm2F0u64cTsECDq4jBNZYhco"
+/// }
+/// ```
+class Credential {
+  Credential({required this.url, required this.token, this.unknownFields});
+
+  /// Create [Credential] instance for bearer tokens.
+  Credential.bearer(this.url, this.token) : unknownFields = null;
+
+  /// Deserialize [json] into [Credential] type.
+  ///
+  /// Throws [FormatException] if [json] is not a valid [Credential].
+  factory Credential.fromJson(Map<String, dynamic> json) {
+    if (json['url'] is! String) {
+      throw FormatException('Url is not provided for the token');
+    }
+
+    final hostedUrl = validateAndNormalizeHostedUrl(json['url'] as String);
+    final token = json['token'] is String ? json['token'] as String : null;
```

Comment:
Let's throw a `FormatException` if `token == null`, then we can remove the nullability...

And if we throw `FormatException` here, that's fine, because in `List<Credential> _loadCredentials() {` we will appropriately log: `Failed to load credentials for ${element['url']}: `